### PR TITLE
Database api endpoints

### PR DESCRIPTION
**Pull request by alexander-e-andrews**
Finished up all I could think of for database API endpoints, will probably go back and comment some more in a future release. There are probably errors in how I return saying there are errors. Please advise with the correct methodology. Also if you try to delete something, and it does not  exist, is that an error. Ask a philosopher.

### CODE REVIEW

File: api/routes/slide.js [hunk 1]
```diff
@@ -14,8 +19,59 @@ router
...
+    try {
+      var newSlide = new Slide()
+      var slideshowID
+      /**
+       * Try CATCH error block
+       * If front end does not send all data, its okay
+       * Only requires a parent slideshow
+       * Catches do nothing, as they are left blank
+       */
+      try {
```
Comment (mirjanneza) [inline api/routes/slide.js]: I'm not sure about all these try catch blocks. What are they needed for?

**Review by mirjanneza — CHANGES_REQUESTED** [review]
There has to be a better way than try catch blocking everything

File: api/routes/slide.js (hunk 1 above)
Comment (wassgha) [reply api/routes/slide.js]: Same concern, I'd rather use `if (typeof req.body.data !== 'undefined')`

File: components/Upload.js [hunk 2]
```diff
@@ -14,7 +14,7 @@ class Upload extends Component {
   handleOnDropAccepted = acceptedFiles => {
     const formData = new FormData()
     formData.append('data', acceptedFiles[acceptedFiles.length - 1])
-    axios.post('/api/slide/upload', formData, {
+    axios.post('/api/v1/slide', formData, {
```
Comment (wassgha) [inline components/Upload.js:19]: This was already fixed, do you have the latest master changes?

File: api/routes/slideshow.js [hunk 3]
```diff
@@ -26,14 +27,47 @@ router
...
         if (!slideshow) return next('Slideshow not found')
-        return slideshow.remove().then(() => {
-          return res.send('success')
-        })
+        /**
+         * return slideshow.remove().then(() => {
+         * return res.send('success')
+         *})
+         */
+        deleteSlides(slideshow.slides)
```
Comment (wassgha) [inline api/routes/slideshow.js]: This needs a return here (as well as inside `deleteSlides`) and need to be able to send `res.send('success')`

File: api/routes/slideshow.js [hunk 4]
```diff
@@ -26,14 +27,47 @@ router
...
     const { id } = req.params
-    return Slideshow.findById(id)
+    return Slideshow.findByIdAndDelete(id)
       .then(slideshow => {
         if (!slideshow) return next('Slideshow not found')
-        return slideshow.remove().then(() => {
-          return res.send('success')
-        })
+        /**
+         * return slideshow.remove().then(() => {
```
Comment (wassgha) [inline api/routes/slideshow.js]: Don't leave code uncommented, just remove it. If there's a problem we'll go back to the git history.

File: api/routes/slide.js [hunk 5]
```diff
@@ -31,37 +87,92 @@ router
...
-})
 
 module.exports = router
+
+/**
+ * This function finds the given slideshow,
+ * then it saves the slide and pushes its id to slideshow
+ * @param {The ID of target slideshow} slideShowID
+ * @param {The slide that is being saved} slide
+ * @param {Whatever next is, ask Wasssseeeeeemmm} next
```
Comment (wassgha) [inline api/routes/slide.js]: Not my name

File: api/routes/slide.js [hunk 6]
```diff
@@ -31,37 +87,92 @@ router
...
+ */
+function addSlideToSlideShow(slideShowID, slide, next) {
+  Slideshow.findById(slideShowID)
```
Comment (wassgha) [inline api/routes/slide.js]: Needs return

File: api/routes/slide.js [hunk 7]
```diff
@@ -31,37 +87,92 @@ router
...
+            .catch(err => next(err))
+        })
+        .catch(err => next(err))
+    })
+    .catch(err => next(err))
+}
+
+function deleteSlideFromSlideShow(slide, next, res) {
+  var slideShowID = slide.slideshow
+  Slideshow.findById(slideShowID).then(slideshow => {
```
Comment (wassgha) [inline api/routes/slide.js]: Needs return

File: api/routes/slide.js [hunk 8]
```diff
@@ -31,37 +87,92 @@ router
...
+ * @param {Whatever next is, ask Wasssseeeeeemmm} next
+ */
+function addSlideToSlideShow(slideShowID, slide, next) {
+  Slideshow.findById(slideShowID)
+    .then(slideshow => {
+      if (!slideshow) return next('Slide not found')
+      slide
+        .save()
+        .then(slide => {
+          if (!slide) return next('Slide not saved ??')
```
Comment (wassgha) [inline api/routes/slide.js]: Are you asking the user??

File: api/routes/slide.js [hunk 9]
```diff
@@ -31,37 +87,92 @@ router
...
+        } catch (e) {}
+        try {
+          slide.description = req.body.description
+        } catch (e) {}
+        try {
+          slide.duration = req.body.duration
+        } catch (e) {}
+        try {
+          slide.slideshow = req.body.slideShowID
+          console.error('Switching slide to a different slideshow currently not implemented')
```
Comment (wassgha) [inline api/routes/slide.js]: and should not be implemented

**Review by wassgha — CHANGES_REQUESTED** [review]
Just a few changes

**Review by mirjanneza — APPROVED** [review]

**Review by wassgha — APPROVED** [review]
LGTM

**Pull request merged**